### PR TITLE
fix: Handle quantized CUDA tensors in async checkpoint writer

### PR DESCRIPTION
**Pull request by sbak5**
# What does this PR do ?
Call dequantize if the tensor has this method. 

This is intended to avoid issues in TE tensors which leads to invalid memory access in the async worker because tensor IPC handle is not created when TE tensors are passed. 

So, we dequantize it explictly until we find a better solution to handle this.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]

File: megatron/core/dist_checkpointing/strategies/filesystem_async.py [hunk 1]
```diff
@@ -154,6 +154,10 @@ def _clone_if_needed(ten: torch.Tensor):
             """
             ten = ten.detach()
             if ten.device.type != "cpu":
+                # We call ``dequantize`` if we detect a quantized tensor on GPU.
+                # This is a workaround to avoid the issue of quantized tensors not being supported by the async writer.
+                if ten.device.type == "cuda" and hasattr(ten, "dequantize"):
```
[the remark this answers is not in the document]
Comment (zhongbozhu) [reply megatron/core/dist_checkpointing/strategies/filesystem_async.py]: Valid comment, @sbak5 please fix this one as well.

**Comment by sbak5** [comment]
@ananthsub will take a look at the Mbridge failure.
